### PR TITLE
feat: add `react-native doctor`

### LINKED ISSUE #51: Health check command
### Introduction

A way to verify react-native installation is complete and working?

### The Core of It

I was just checking out Flutter & I see there is a thing there called `flutter doctor` which ensures that installation is complete & working. I think it makes sense for React Native to have something like a `react-native doctor` or `react-native health` or something else.

### Discussion points

Might be beneficial for beginners who have a lot of trouble installing Android Studio if they are on anything else than a Mac

### Example of Flutter Doctor

Here's an output of the command `flutter doctor -v` in the terminal -

```bash
[✓] Flutter (Channel master, v1.1.2-pre.39, on Mac OS X 10.14.1 18B75, locale en-IN)
    • Flutter version 1.1.2-pre.39 at /Volumes/Coding/Code/Flutter/installation/flutter
    • Framework revision 0e9ad43416 (34 hours ago), 2018-12-16 16:39:28 -0500
    • Engine revision 4941125829
    • Dart version 2.2.0 (build 2.2.0-dev.1.1 f9ebf21297)

[✓] Android toolchain - develop for Android devices (Android SDK 28.0.3)
    • Android SDK at /Users/deadcoder0904/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 28.0.3
    • ANDROID_HOME = /Users/deadcoder0904/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 10.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.1, Build version 10B61
    • ios-deploy 1.9.4
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 31.1.1
    • Dart plugin version 181.5656
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)

[✓] VS Code (version 1.30.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 2.21.1

[!] Connected device
    ! No devices available

! Doctor found issues in 1 category.
```

This issue has been moved from https://github.com/react-native-community/discussions-and-proposals/issues/73

### PR DESCRIPTION
**Pull request by lucasbento**
Summary:
---------

This PR fixes #51, it adds a `doctor` command to verify the softwares/libraries/tools that we depend on and make sure that they match the minimum requirements needed by react-native.

The version for `Node.js` is not correct as I put it there for testing purposes.

---

react-native doctor command

Test Plan:
----------

1. Uninstall `cocoapods` or `ios-deploy` (easiest ones to try);
1. Run `doctor` command

TODOS:
----------

- [x] List number of problems & warnings
- [x] Separate list of issues in a different file and each issue should have their own file to run the automatic fix
- [x] Add automatic fix for the softwares
- [x] For "no automatically fix" softwares there should be a link to download new versions;
- [x] Add prompt question to fix after it has been ran
- [x] Add `--fix` argument that should go directly to fixing it
- [ ] Add `--dry-run` option
- [x] Add `--contributor` option to only show Android NDK if specified
- [ ] Make sure that the version ranges are good to go
- [x] Add more softwares/tools/libs

### CODE REVIEW

**Comment by patrickkempff** [comment]
Wow, this looks awesome!

**Comment by Inovassist-dev** [comment]
Oh my god, I must say, we need that!
Flutter has its own doctor command and it's really good. I wonder how many bugs and bad environment configs could be solved with an on-the-fly "doctor" command!

**Comment by lucasbento** [comment]
@Inovassist-dev: thanks for the feedback on this, that's what we are aiming for!

This week I'm pretty busy, gonna get back to work on this by next week 🙂

**Comment by rickhanlonii** [comment]
Hey @lucasbento, this is amazing! I'd love to help you get this through, let me know what you need!

**Comment by sahrens** [comment]
This would be great, thanks for getting it started!

Thoughts on building this in such a way that we can compose it with other tools and IDEs (e.g. support JSON output instead of logging direct to console). cc @thymikee

**Comment by cpojer** [comment]
Can we call this `react-native fix`?

**Comment by Inovassist-dev** [comment]
I don't think that fix would be a great idea.
Fluter already has the doctor command and, sincerely, it's a big deal.
As it's not going to necessarily fix React installation, we can call it doctor or repair, just to chase possible errors and, if found, fix them

**Comment by thymikee** [comment]
@sahrens definitely open to that

**Comment by TheSavior** [comment]
I'd love to see this landed and iterated upon. How many features do we need to wait for this to have before landing an MVP?

**Comment by lucasbento** [comment]
Those are the installations that _I would like to have in_:

##### Platform agnostic

- Current folder is a `react-native` project (check if `react-native` is installed)
- `yarn` or `npm` are in a recent version (need to store the minimum version)
- Minimum node version (need to store the minimum version)
- Check watchman (need to store the minimum version)

##### Android

- Check if `ANDROID_HOME` is defined
- Android SDK
- Android NDK (consider not showing every time as it's not needed for everyone)

##### iOS

- Check Xcode
- Check CocoaPods
- Check if `ios-deploy` is installed (need to tell the user somehow that this is only needed with real devices)

Totally discussable though, I would like to have this on `master` as soon as possible so it's easier for people to add more stuff.

@rickhanlonii @sahrens: a review on the source would be very much appreciated, I'm trying to build something as modular as possible, the JSON output was already asked on #51 and I'm definitely on board with that.

The first thing that I need help with is a little review and adding more installation checks, if anyone of you want to help with that then let's do it!

**Comment by tabrindle** [comment]
I realize that you are looking to add more functionality in the future, but right now this is exactly the feature set of Solidarity https://github.com/infinitered/solidarity Might we consider using this and combining efforts on the automatic fix part on the wishlist?

**Comment by rickhanlonii** [comment]
@tabrindle Aolidariy is great! Will support things like checking for IDE installations, android SDKs, xcode versions, vscode/intellij/android studio plugins?

File: packages/cli/src/commands/doctor/checkInstallation.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    await commandExists(command);
+
+    return true;
+  } catch (_ignored) {
+    return false;
+  }
+};
+
+const doesSoftwareNeedToBeFixed = ({version, versionRange}) =>
+  version === 'Not Found' || !semver.satisfies(version, versionRange);
```
Comment (rickhanlonii) [inline packages/cli/src/commands/doctor/checkInstallation.js]: Note that semver can throw here if for some reason one of the versions are invalid. e.g. for me, I'm getting the string "Using globally installed version of Yarn"

Screenshot 2019-08-07 11 33 17

We'll at least need to try/catch here, but in the catch we may want to log a quality error and consider crashing because if we don't get a semver by this point that the individual issue needs to diagnose why they don't have a semver

File: packages/cli/src/commands/doctor/doctor.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
+import chalk from 'chalk';
+import envinfo from 'envinfo';
+import {logger} from '@react-native-community/cli-tools';
+import ISSUES from './issues';
```
Comment (rickhanlonii) [inline packages/cli/src/commands/doctor/doctor.js]: Reading through these changes, I struggled with the term "issue" because until these are checked there is nothing wrong (i.e. the issues are not issues)

What do you think of calling these something like healthchecks?

File: packages/cli/src/commands/doctor/doctor.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+      ? chalk.red('✖')
+      : chalk.yellow('●')
+    : chalk.green('✓');
+
+  logger.log(` ${symbol} ${label}`);
+};
+
+export default (async function runDoctor() {
+  const Loader = getLoader();
+  const loader = new Loader();
```
Comment (rickhanlonii) [inline packages/cli/src/commands/doctor/doctor.js]: Nit: naming this a "loader" when it's not really showing a loading message and is mostly only a logger is hard to read (I recognize that this is called a loader elsewhere so we should either change all usages or stay consistent)

File: packages/cli/src/commands/doctor/doctor.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        npmPackages: ['react', 'react-native', '@react-native-community/cli'],
+        npmGlobalPackages: ['*react-native*'],
+      },
+      {json: true, showNotFound: true},
+    ),
+  );
+
+  const iterateOverIssues = async ({label, issues}) => ({
+    label,
+    issues: (await Promise.all(
```
Comment (rickhanlonii) [inline packages/cli/src/commands/doctor/doctor.js]: Nice use of Promise.all to run all the checks at the same time, this is heavily I/O bound so this should be helpful

File: packages/cli/src/commands/doctor/doctor.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        return stats.warnings++;
+      }
+
+      if (issue.needsToBeFixed) {
+        return stats.errors++;
+      }
+    });
+  });
+
+  // Print overall stats
```
Comment (rickhanlonii) [inline packages/cli/src/commands/doctor/doctor.js]: Consider moving to a `printStats` function instead of using a comment

File: packages/cli/src/commands/doctor/doctor.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+        const {needsToBeFixed} = issue.getDiagnostics
+          ? issue.getDiagnostics(environmentInfo)
+          : await issue.getDiagnosticsAsync(environmentInfo);
+
+        return {
+          label: issue.label,
+          needsToBeFixed,
+          runAutomaticFix: issue.runAutomaticFix,
+          // Assume that it's required unless specified otherwise
```
Comment (rickhanlonii) [inline packages/cli/src/commands/doctor/doctor.js]: Comment probably isn't needed

File: packages/cli/src/commands/doctor/runAutomaticFix.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    for (const issueToFix of issuesToFix) {
+      const issueSpinner = ora({
+        prefixText: '',
+        text: issueToFix.label,
+      }).start();
+
+      try {
+        await issueToFix.runAutomaticFix({loader: issueSpinner});
+      } catch (error) {
+        // TODO: log the error in a meaningful way
```
Comment (rickhanlonii) [inline packages/cli/src/commands/doctor/runAutomaticFix.js]: Would console.error be better than nothing for now?

**Comment by tabrindle** [comment]
Solidarity can do anything envinfo does, as it uses it under the hood.

File: packages/cli/src/commands/doctor/doctor.js [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+      }),
+    )).filter(issue => !!issue),
+  });
+
+  const iterateOverCategories = categories =>
+    Promise.all(categories.map(iterateOverIssues));
+
+  const issuesPerCategory = await iterateOverCategories(Object.values(ISSUES));
+
+  loader.stop();
```
Comment (rickhanlonii) [inline packages/cli/src/commands/doctor/doctor.js]: Note that the OraMock does not have a stop function, causing:

Screenshot 2019-08-07 11 49 45

**Comment by Inovassist-dev** [comment]
@tabrindle, I think it would be nice to have something like Solidarity already integrated with React Native, dont you think?

**Comment by lucasbento** [comment]
@rickhanlonii: thank you so much for the review! gonna check it out.

> I realize that you are looking to add more functionality in the future, but right now this is exactly the feature set of Solidarity infinitered/solidarity Might we consider using this and combining efforts on the automatic fix part on the wishlist?

@tabrindle: I checked solidarity before starting this and I struggled to understand how it would be used on the core of the CLI, perhaps we should discuss its possibilities somewhere else?

File: packages/cli/src/commands/doctor/doctor.js (hunk 2 above)
Comment (lucasbento) [reply packages/cli/src/commands/doctor/doctor.js]: I also struggled to name these, definitely changing it to `healthchecks`!

File: packages/cli/src/commands/doctor/runAutomaticFix.js (hunk 7 above)
Comment (lucasbento) [reply packages/cli/src/commands/doctor/runAutomaticFix.js]: Didn't pay much attention to this part yet, it's hard to log an error right now as the loader is on top of everything, so first it needs to stop the loader, log an error and then continue to the next automatic fix.

File: packages/cli/src/commands/doctor/doctor.js (hunk 2 above)
Comment (lucasbento) [reply packages/cli/src/commands/doctor/doctor.js]: I also struggled to name these, definitely changing it to healthchecks!

File: packages/cli/src/commands/doctor/doctor.js (hunk 8 above)
Comment (lucasbento) [reply packages/cli/src/commands/doctor/doctor.js]: I wonder how you got this as I don't have it locally.

It does have a `stop` method: https://github.com/sindresorhus/ora#stop.

File: packages/cli/src/commands/doctor/checkInstallation.js (hunk 1 above)
Comment (lucasbento) [reply packages/cli/src/commands/doctor/checkInstallation.js]: How did you install yarn? I want to reproduce that locally.

File: packages/cli/src/commands/doctor/doctor.js (hunk 8 above)
Comment (lucasbento) [reply packages/cli/src/commands/doctor/doctor.js]: I got it now by running with `--verbose`, I'll check that out, thanks!

File: packages/cli/src/commands/doctor/checkInstallation.js (hunk 1 above)
Comment (rickhanlonii) [reply packages/cli/src/commands/doctor/checkInstallation.js]: Probably `npm install -g yarn`

**Comment by rickhanlonii** [comment]
@lucasbento do you want to get this in an start iterating?

**Comment by thymikee** [comment]
I like the direction where it's going. Let's make sure to mark it as experimental (e.g. in the description and the docs, or add docs later) and it doesn't break stable commands. And ship it :shipit:

File: packages/cli/src/commands/doctor/versionRanges.js [hunk 9]
```diff
@@ -0,0 +1,12 @@
+export default {
+  // Common
+  NODE_JS: '>= 11 <= 12', //'10.x',
```
Comment (tido64) [inline packages/cli/src/commands/doctor/versionRanges.js]: I'm confused. Are we supposed to allow 10.x here? Otherwise, this looks like it only allows 11.x and up.

Comment (thymikee) [reply packages/cli/src/commands/doctor/versionRanges.js]: Node.js should be `>= 8.3`

Comment (lucasbento) [reply packages/cli/src/commands/doctor/versionRanges.js]: The versions haven't been yet checked against what should be supported.

This version for `Node.js` is merely for testing.

File: packages/cli/src/commands/doctor/healthchecks/cocoaPods.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+const cocoaPods = {
+  label: 'CocoaPods',
+  getDiagnosticsAsync: async () => ({
+    needsToBeFixed: !(await isSoftwareInstalled('pod')),
+  }),
+  runAutomaticFix: async ({loader}) => {
+    try {
+      // First attempt to install `cocoapods`
+      await execa('gem', ['install', 'cocoapods']);
```
Comment (tido64) [inline packages/cli/src/commands/doctor/healthchecks/cocoaPods.js]: Should we rely on Homebrew/MacPorts to avoid having to invoke `sudo`? Or at least use them if they are installed.

File: packages/cli/src/commands/doctor/healthchecks/androidHomeEnvVariable.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+// List of answers on how to set `ANDROID_HOME` for each platform
+const URLS = {
+  darwin: 'https://stackoverflow.com/a/28296325/4252781',
+  win32: 'https://stackoverflow.com/a/54888107/4252781',
+  linux: 'https://stackoverflow.com/a/39228100/4252781',
+};
+
+const label = 'ANDROID_HOME';
+
+const iosDeploy = {
```
Comment (sweggersen) [inline packages/cli/src/commands/doctor/healthchecks/androidHomeEnvVariable.js]: rename

File: packages/cli/src/commands/doctor/healthchecks/androidNDK.js [hunk 12]
```diff
@@ -0,0 +1,35 @@
+import chalk from 'chalk';
+import {logManualInstallation} from './common';
+import versionRanges from '../versionRanges';
+import {doesSoftwareNeedToBeFixed} from '../checkInstallation';
+
+const iosDeploy = {
```
Comment (sweggersen) [inline packages/cli/src/commands/doctor/healthchecks/androidNDK.js]: rename

File: packages/cli/src/commands/doctor/healthchecks/androidSDK.js [hunk 13]
```diff
@@ -0,0 +1,35 @@
+import chalk from 'chalk';
+import {logManualInstallation} from './common';
+import versionRanges from '../versionRanges';
+import {doesSoftwareNeedToBeFixed} from '../checkInstallation';
+
+const iosDeploy = {
```
Comment (sweggersen) [inline packages/cli/src/commands/doctor/healthchecks/androidSDK.js]: rename

File: packages/cli/src/commands/doctor/doctor.js [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+          healthchecks: removeFixedCategories(healthchecksPerCategory),
+          automaticFixLevel: automaticFixLevel[key],
+          stats,
+          loader,
+          environmentInfo,
+        });
+
+        process.exit(0);
+      } catch (err) {
+        // TODO: log error
```
Comment (tido64) [inline packages/cli/src/commands/doctor/doctor.js]: Is this going to be done now or in a separate PR?

File: packages/cli/src/commands/doctor/healthchecks/androidHomeEnvVariable.js [hunk 15]
```diff
@@ -0,0 +1,29 @@
+import chalk from 'chalk';
+import {logManualInstallation} from './common';
+
+// List of answers on how to set `ANDROID_HOME` for each platform
+const URLS = {
+  darwin: 'https://stackoverflow.com/a/28296325/4252781',
```
Comment (sweggersen) [inline packages/cli/src/commands/doctor/healthchecks/androidHomeEnvVariable.js]: Maybe it would be smart to inline these answers

File: packages/cli/src/commands/doctor/healthchecks/packageManagers.js [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    needsToBeFixed: doesSoftwareNeedToBeFixed({
+      version: Binaries.Yarn.version,
+      versionRange: versionRanges.YARN,
+    }),
+  }),
+  // Only show `yarn` if there's a `yarn.lock` file in the current directory
+  // or if we can't identify that the user uses yarn or npm
+  visible:
+    packageManager === PACKAGE_MANAGERS.YARN || packageManager === undefined,
+  runAutomaticFix: () => console.log('should fix node'),
```
Comment (tido64) [inline packages/cli/src/commands/doctor/healthchecks/packageManagers.js]: Is this intentional? Isn't Yarn installed separately?

File: packages/cli/src/commands/doctor/healthchecks/watchman.js [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+
+const watchman = {
+  label: 'Watchman',
+  getDiagnostics: ({Binaries}) => ({
+    needsToBeFixed: doesSoftwareNeedToBeFixed({
+      version: Binaries.Watchman.version,
+      versionRange: versionRanges.WATCHMAN,
+    }),
+  }),
+  runAutomaticFix: () => console.log('should fix watchman'),
```
Comment (tido64) [inline packages/cli/src/commands/doctor/healthchecks/watchman.js]: Is this placeholder text?

File: packages/cli/src/commands/doctor/healthchecks/cocoaPods.js (hunk 10 above)
Comment (thymikee) [reply packages/cli/src/commands/doctor/healthchecks/cocoaPods.js]: we have a helper function for that, let's use it :) https://github.com/react-native-community/cli/blob/master/packages/cli/src/tools/installPods.js

File: packages/cli/src/commands/doctor/doctor.js (hunk 8 above)
Comment (sweggersen) [reply packages/cli/src/commands/doctor/doctor.js]: If you look at tools/loader.js the `getLoader` returns a `OraMock` if `verbose`. I guess we could just add a `stop` stub to `OraMock`
